taskcat/template_params.py

```diff
@@ -0,0 +1,361 @@
+
+
+    @staticmethod
+    def regxfind(re_object, data_line):
+        """
+        Returns the matching string.
+
+        :param re_object: Regex object
+        :param data_line: String to be searched
+
+        :return: Matching String if found, otherwise return 'Not-found'
+        """
+        sg = re_object.search(data_line)
+        if sg:
+            return str(sg.group())
+        else:
+            return str('Not-found')
+
+    def get_available_azs(self, count):
+        """
+        Returns a list of availability zones in a given region.
+
+        :param region: Region for the availability zones
+        :param count: Minimum number of availability zones needed
+
+        :return: List of availability zones in a given region
+
+        """
+        ec2_client = self._boto_client.get('ec2', region=self.region)
+        available_azs = []
+        availability_zones = ec2_client.describe_availability_zones(
+            Filters=[{'Name': 'state', 'Values': ['available']}])
+
+        for az in availability_zones['AvailabilityZones']:
+            available_azs.append(az['ZoneName'])
+
+        if len(available_azs) < count:
+            print("{0}!Only {1} az's are available in {2}".format(PrintMsg.ERROR, len(available_azs), self.region))
+            raise TaskCatException
+        else:
+            azs = ','.join(available_azs[:count])
+            return azs
+
+    def get_single_az(self, az_id):
+        """
+        Get a single valid AZ for the region.
+        The number passed indicates the ordinal representing the AZ returned.
+        For instance, in the 'us-east-1' region, providing '1' as the ID would
+        return 'us-east-1a', providing '2' would return 'us-east-1b', etc.
+        In this way it's possible to get availability zones that are
+        guaranteed to be different without knowing their names.
+        :param region: Region of the availability zone
+        :param az_id: 0-based ordinal of the AZ to get
+        :return: The requested availability zone of the specified region.
+        """
+
+        regional_azs = self.get_available_azs(az_id)
+        return regional_azs.split(',')[-1]
+
+    def get_content(self, bucket, object_key):
+        """
+        Returns the content of an object, given the bucket name and the key of the object
+
+        :param bucket: Bucket name
+        :param object_key: Key of the object
+        :param object_key: Key of the object
+        :return: Content of the object
+
+        """
+        s3_client = self._boto_client.get('s3', region=self.region, s3v4=True)
+        try:
+            dict_object = s3_client.get_object(Bucket=bucket, Key=object_key)
+        except TaskCatException:
+            raise
+        except Exception:
+            print("{} Attempted to fetch Bucket: {}, Key: {}".format(PrintMsg.ERROR, bucket, object_key))
+            raise
+        content = dict_object['Body'].read().decode('utf-8').strip()
+        return content
```

Comment:
Looking at `ClientFactory.get` it too can raise TaskCatException, so it'd be good to place inside the try.

Also, given that the `return` statements are only executed if there are no exceptions, these should also be inside the `try` block. This prevents any dangling logic and clarifies the execution flow:
```
def my_function(params):
  try:
    result = compute_some_stuff_or_get_stuff
    the_thing_that_matters = extract_thing_from_result
    return the_thing_that_matters
  except SomeException:
    raise
```
Reading this I can see what the happy path is right away and then move on to the not-happy path by seeing how errors are handled.

On a related note:
1. if we are propagating the exceptions back up to the callee... does it matter to handle them?
2. if we are handling them for logging purposes, shouldn't we log the TaskCatException as well?
3. if we are going to log the errors, should we also log the message from the exception (maybe even as a debug or trace?)

Reply:
that's fair. We can get this merged and then look into a future change that tackles this bit in isolation.

would like to note that your concern is in itself a good argument to look into this method further.